### PR TITLE
Fablup/attack of the pickle

### PR DESCRIPTION
**Pull request by ggorman**

### CODE REVIEW

File: tests/test_pickle.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import numpy as np
+from examples.seismic import demo_model
+from examples.seismic.source import TimeAxis, RickerSource
+from devito.dimension import TimeDimension, SteppingDimension
+from devito.base import Constant
+
+import cloudpickle as pickle
+
+
+def test_pickle():
```
Comment (FabioLuporini) [inline tests/test_pickle.py]: this needs to be split into multiple smaller tests

docstrings should be added

File: tests/test_pickle.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    ricker = RickerSource(name='src', grid=model.grid, f0=f0, time_range=time_range)
+
+    pkl_ricker = pickle.dumps(ricker)
+    new_ricker = pickle.loads(pkl_ricker)
+    assert np.isclose(np.linalg.norm(ricker.data.flatten()),
+                      np.linalg.norm(new_ricker.data.flatten()))
+
+
+if __name__ == "__main__":
```
Comment (FabioLuporini) [inline tests/test_pickle.py]: useless

**Comment by FabioLuporini** [comment]
we also need to figure out how to retrieve the base pointer of `Function._data` upon unpickling

one possibility emerged during offline discussion is to provide a placeholder object to let the receiver fetch the data directly from disk ()

**Comment by FabioLuporini** [comment]
one last thing: this might require further coding to work with YASK

File: devito/function.py [hunk 3]
```diff
@@ -566,6 +566,9 @@ def laplace2(self, weight=1):
         return sum([second_derivative(first * weight, dim=d, order=order)
                     for d in self.space_dimensions])
 
+    # Pickling support
+    _pickle_kwargs = TensorFunction._pickle_kwargs + ['grid']
+
```
Comment (tjb900) [inline devito/function.py]: Presumably `shape` and `dimensions` too?  What about space order etc?

File: devito/tools/abc.py [hunk 4]
```diff
@@ -84,3 +84,38 @@ def _signature_items(self):
...
+
+    _pickle_args = []
+    """The positional arguments that need to be passed to __new__ upon unpickling."""
+
+    _pickle_kwargs = []
+    """The keyword arguments that need to be passed to __new__ upon unpickling."""
+
+    def __reduce_ex__(self, proto):
+        return object.__reduce_ex__(self, proto)
+
```
Comment (tjb900) [inline devito/tools/abc.py]: Unless we are expecting to need to add functionality (or debugging), `__reduce_ex__ = object.__reduce_ex__` ?

File: devito/function.py [hunk 5]
```diff
@@ -566,6 +566,9 @@ def laplace2(self, weight=1):
...
 
 class TimeFunction(Function):
     """
```
Comment (tjb900) [inline devito/function.py]: Will probably need to add `time_order`, `time_dim`, etc here also, I guess?

Comment (FabioLuporini) [reply devito/function.py]: think so. @ggorman ?

**Comment by FabioLuporini** [comment]
Merged.

**Pull request merged**